bigbluebutton-html5/imports/ui/components/video-dock/component.jsx

```diff
@@ -1,11 +1,325 @@
+    const ws = this.state.ws;
+    const { users } = this.props;
+    for (let i = 0; i < users.length; i++) {
+      if (users[i].has_stream) {
+        this.start(users[i].userId, false, this.refs.videoInput);
+      }
+    }
+
+    document.addEventListener('joinVideo', () => { that.shareWebcam(); });// TODO find a better way to do this
+    document.addEventListener('exitVideo', () => { that.unshareWebcam(); });
+
+    ws.addEventListener('message', (msg) => {
+      const parsedMessage = JSON.parse(msg.data);
+
+      console.log('Received message new ws message: ');
+      console.log(parsedMessage);
+
+      switch (parsedMessage.id) {
+
+        case 'startResponse':
+          this.startResponse(parsedMessage);
+          break;
+
+        case 'error':
+          this.handleError(parsedMessage);
+          break;
+
+        case 'playStart':
+          this.handlePlayStart(parsedMessage);
+          break;
+
+        case 'playStop':
+          this.handlePlayStop(parsedMessage);
+
+          break;
+
+        case 'iceCandidate':
+
+          const webRtcPeer = this.state.webRtcPeers[parsedMessage.cameraId];
+
+          if (webRtcPeer !== null) {
+            webRtcPeer.addIceCandidate(parsedMessage.candidate, (err) => {
+              if (err) {
+                console.error(`Error adding candidate: ${err}`);
+              }
+            });
+          } else {
+            log('error', ' [ICE] Message arrived before webRtcPeer?');
+          }
+
+          break;
+
+      }
+    });
+  }
+
+  start(id, shareWebcam, videoInput) {
+    const that = this;
+
+    const ws = this.state.ws;
+
+    console.log(`Starting video call for video: ${id}`);
+    log('info', 'Creating WebRtcPeer and generating local sdp offer ...');
+
+    const onIceCandidate = function (candidate) {
+      const message = {
+        type: 'video',
+        role: shareWebcam? 'share' : 'viewer',
+        id: 'onIceCandidate',
+        candidate,
+        cameraId: id,
+      };
+      that.sendMessage(message);
+    };
+
+    const options = {
+      mediaConstraints: { audio: false,
+        video: {
+          width: {min: 320, ideal: 320},
+          height: {min: 240, ideal:240},
+          frameRate: { min: 5, ideal: 10}
+        }
+      },
+      onicecandidate: onIceCandidate,
+    };
+
+    let peerObj;
+    if (shareWebcam) {
+      options.localVideo = videoInput;
+      peerObj = kurentoUtils.WebRtcPeer.WebRtcPeerSendonly;
+    } else {
+      peerObj = kurentoUtils.WebRtcPeer.WebRtcPeerRecvonly;
+
+      options.remoteVideo = document.createElement('video');
+      options.remoteVideo.id = `video-elem-${id}`;
+      options.remoteVideo.width = 120;
+      options.remoteVideo.height = 90;
+      options.remoteVideo.autoplay = true;
+      options.remoteVideo.playsinline = true;
+
+      document.getElementById('webcamArea').appendChild(options.remoteVideo);
+    }
+
+    this.state.webRtcPeers[id] = new peerObj(options, function (error) {
```

Comment:
May I ask - you do realize that changing the state variables like this won't trigger an update (or re-render)?
(Same for the line below: `this.state.wsQueue.push(message);` )

If you don't need to trigger a React lifecycle update and you just need to save some stuff - then you don't have to use state for that, you can simply use `this`.

For example:

```
constructor(props) {
  super(props);
  this.state = {
    videos: {}
    ...
  };
  this.webRtcPeers = {};
  this.wsQueue = [];
  ...
};
```
And here you could refer to it `this.webRtcPeers[id] = ..` instead of `this.state.webRtcPeers[id]`